zephyr/CMakeLists.txt

```
@@ -1,187 +1 @@
-    sign
-    --key ${MCUBOOT_BASE}/${CONFIG_BOOT_SIGNATURE_KEY_FILE}
-    --header-size $<TARGET_PROPERTY:partition_manager,PM_MCUBOOT_PAD_SIZE>
-    --align       ${CONFIG_DT_FLASH_WRITE_BLOCK_SIZE}
-    --version     ${CONFIG_MCUBOOT_IMAGE_VERSION}
-    --slot-size   $<TARGET_PROPERTY:partition_manager,PM_MCUBOOT_PRIMARY_SIZE>
-    --pad-header
-    )
-
-  set(app_offset $<TARGET_PROPERTY:partition_manager,PM_MCUBOOT_PRIMARY_SIZE>)
-
-  sign(${app_to_sign_hex}     # Hex to sign
-    ${PROJECT_BINARY_DIR}/app # Prefix for generated files
-    ${app_offset}             # Offset
-    ${app_sign_depends}       # Dependencies
-    app_signed_hex            # Generated hex output variable
-    )
-
-  add_custom_target(mcuboot_sign_target DEPENDS ${app_signed_hex})
-
-  set_property(GLOBAL PROPERTY
-    mcuboot_primary_app_PM_HEX_FILE
-    ${app_signed_hex}
-    )
-  set_property(GLOBAL PROPERTY
-    mcuboot_primary_app_PM_TARGET
-    mcuboot_sign_target
-    )
-
-  if (CONFIG_MCUBOOT_BUILD_S1_VARIANT)
-    # Secure Boot (B0) is enabled, and we have to build update candidates
-    # for both S1 and S0.
-
-    # We need to override some attributes of the parent slot S0/S1.
-    # Which contains both the S0/S1 image and the padding/header.
-    foreach(parent_slot s0;s1)
-      set(slot ${parent_slot}_image)
-
-      # Fetch the target and hex file for the current slot.
-      # Note that these hex files are already signed by B0.
-      get_property(${slot}_target GLOBAL PROPERTY ${slot}_PM_TARGET)
-      get_property(${slot}_hex GLOBAL PROPERTY ${slot}_PM_HEX_FILE)
-
-      # The gap from S0/S1 partition is calculated by partition manager
-      # and stored in its target.
-      set(slot_offset
-        $<TARGET_PROPERTY:partition_manager,${parent_slot}_TO_SECONDARY>)
-
-      # Depend on both the target for the hex file, and the hex file itself.
-      set(dependencies "${${slot}_target};${${slot}_hex}")
-
-      set(out_path ${PROJECT_BINARY_DIR}/signed_by_mcuboot_and_b0_${slot})
-
-      sign(${${slot}_hex} # Hex file to sign
-        ${out_path}
-        ${slot_offset}
-        "${dependencies}" # Need "..." to make it a list.
-        signed_hex        # Created file variable
-        )
-
-      # We now have to override the S0/S1 partition, so use `parent_slot`
-      # variable, which is "s0" and "s1" respectively. This to get partition
-      # manager to override the implicitly assigned container hex files.
-
-      # Wrapper target for the generated hex file.
-      add_custom_target(signed_${parent_slot}_target DEPENDS ${signed_hex})
-
-      # Override the container hex file.
-      set_property(GLOBAL PROPERTY
-        ${parent_slot}_PM_HEX_FILE
-        ${signed_hex}
-        )
-
-      # Override the container hex file target.
-      set_property(GLOBAL PROPERTY
-        ${parent_slot}_PM_TARGET
-        signed_${parent_slot}_target
-        )
-    endforeach()
-  endif()
-endif()
+include(../../nrf/cmake/mcuboot.cmake)
```

Sorry for the slow response -- last week there were 2 US holidays. Thanks a lot for working on this.

Can you explain why this needs to remain here? I naively assumed that invoking this logic could be done in the app/CMakeLists.txt for the root image, since I think that's the one that ultimately decides the boot chain and therefore what needs signing.

Not trying to get in the way, just trying to understand better.

Thanks a lot!

---

No problem :) 

AFAIK this is the piece of code which is executed when this repo is included as a module. 
It made sense to be to have this here as it seems that this file is executed automatically.

---

> No problem :)

:)
 
> AFAIK this is the piece of code which is executed when this repo is included as a module.

Yup, that's right.

> It made sense to be to have this here as it seems that this file is executed automatically.

I'm still confused about this issue from my first comment, sorry:

> I naively assumed that invoking this logic could be done in the app/CMakeLists.txt for the root image, since I think that's the one that ultimately decides the boot chain and therefore what needs signing.

---

> I'm still confused about this issue from my first comment, sorry:
> 
> > I naively assumed that invoking this logic could be done in the app/CMakeLists.txt for the root image, since I think that's the one that ultimately decides the boot chain and therefore what needs signing.

In that case, each application must include some boilerplate *if* it is going to be used with MCUBoot. Even if its only a single line of `include` I much prefer the solution proposed in this PR.

---

> In that case, each application must include some boilerplate _if_ it is going to be used with MCUBoot. Even if its only a single line of `include` I much prefer the solution proposed in this PR.

Summary of what we discussed offline:

1. we'll go with this for now
2. we can try moving this content into the module CMakeLists.txt for the nrf repository, instead of its current location (module CMakeLists.txt for the mcuboot repository, which may not be needed at all as a file after the new multi-image implementation goes in)

Did I get that right?

---

Yes!

---

```suggestion
include(${ZEPHYR_NRF_MODULE_DIR})
```
cause I believe this is being executed as part of Zephyr module inclusion

---

On second thought.
Any special reason this include needs to happen here ?
Could this `include` code be moved here: https://github.com/NordicPlayground/fw-nrfconnect-nrf/blob/master/CMakeLists.txt#L10-L19
instead.